### PR TITLE
Actions in redux

### PR DESCRIPTION
**Pull request by PeteCoward**
requires canoe-backend https://github.com/catalpainternational/canoe-backend/pull/148

- [ ] clicking through all the cards in a lesson, sets that lesson to complete
   - [ ] answers are stored with an attempt count
- [ ] finishing the last lesson in a course completes the course too if there is no exam
- [ ] completing a pre-learning exam completes the course
- [ ] finishing a course exam but failing, does not complete the course
- [ ] passing a course exam does complete the course 
- [ ] logging out clears state ( you go the login page so can't really tell )
  - [ ] without refresh logging in as a new account should not show you old data
  - [ ] logging back in as the same account should restore old data

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ subscribeToStore(() => {
         case "update-waiting":
             // TODO should we reload? it might interrupt something
             // should we prompt the user?
-            window.location.reload(true);
+            window.location.reload();
```
Comment (PeteCoward) [inline src/index.js:47]: passing true is a deprecated interface, not sure why it was there to begin with

File: src/riot/Lesson/TestMultipleAnswer.riot.html [hunk 2]
```diff
@@ -47,107 +48,94 @@ <h4 if="{state.visibleAnswers.length > 0}">{ TRANSLATIONS.checkCorrect() }</h4>
...
-                let { visibleAnswers, hiddenAnswers } = this.state;
-
-                if (this.doesArrayIncludeAnswer(visibleAnswers, answer)) {
-                    visibleAnswers = visibleAnswers.filter(
-                        visible => !doAnswersMatch(visible, answer)
-                    );
-                } else if (this.doesArrayIncludeAnswer(hiddenAnswers, answer)) {
-                    hiddenAnswers = hiddenAnswers.filter(hidden => !doAnswersMatch(hidden, answer));
+            selectAnswer(answer) {
+                const storedAnswerData = this.getAnswer();
```
Comment (PeteCoward) [inline src/riot/Lesson/TestMultipleAnswer.riot.html:117]: this is a candidate to move into the page ts code

File: src/ts/Implementations/Specific/Course.ts [hunk 3]
```diff
@@ -1,72 +1,133 @@
...
-    getExamHighScore as getHighScore,
-    hasUserTriedExam as hasTriedExam,
-} from "js/actions/exam";
-import ExamGrader from "js/ExamGrader";
+    getExamScore,
+    storeExamScore,
+    getTestAnswers,
+    clearPageTestAnswers,
+} from "ReduxImpl/Interface";
+import { persistExamScore } from "js/actions/ExamScores";
```
Comment (md8n) [inline src/ts/Implementations/Specific/Course.ts:8]: You need to add a typings file for this

**Review by md8n — CHANGES_REQUESTED** [review]

File: src/ts/Implementations/Specific/Course.ts (hunk 3 above)
Comment (PeteCoward) [reply src/ts/Implementations/Specific/Course.ts:8]: It's there, I had not pushed my latest

**Comment by PeteCoward** [comment]
I pushed an error!

File: src/ts/Implementations/Specific/Course.ts (hunk 3 above)
Comment (md8n) [reply src/ts/Implementations/Specific/Course.ts:8]: You're still missing that typings file

**Review by md8n — APPROVED** [review]

File: src/ts/Implementations/Specific/Lesson.ts [hunk 4]
```diff
@@ -1,30 +1,41 @@
...
     get shortDescription(): string {
         return this.storedData?.description;
     }
     get longDescription(): string {
         return this.storedData?.long_description;
     }
-    get content(): any {
-        return this.storedData?.content;
+    get cards(): any {
+        return this.storedData?.cards;
```
Comment (megganeturner) [inline src/ts/Implementations/Specific/Lesson.ts:16]: cards are nested in content, so this should be `return this.storedData?.content.cards;`

Comment (PeteCoward) [reply src/ts/Implementations/Specific/Lesson.ts:16]: not if you are running the accompanying backend pr catalpainternational/canoe-backend#148

Comment (megganeturner) [reply src/ts/Implementations/Specific/Lesson.ts:16]: Ah! 😱 Let me check all my other comments are still relevant 🤦🏼‍♀️

Comment (megganeturner) [reply src/ts/Implementations/Specific/Lesson.ts:16]: This one is fine, but everything else still stands (:

File: src/riot/Components/Card.riot.html [hunk 5]
```diff
@@ -23,7 +22,24 @@ <h5 class="clamp-lines">{ props.title || props.contentItem.title }</h5>
         export default {
             getCardImage() {
                 return getCardImageUrl(this.props.link || this.props.contentItem.id, this.props.imageUrl);
```
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/Components/Card.riot.html]: see [Fix card images](https://github.com/catalpainternational/canoe/pull/598/commits/f4c30b5110f524a6bd1353e7c5670182e4c821ff)

File: src/riot/Lesson/ExamResults.riot.html [hunk 6]
```diff
@@ -1,42 +1,35 @@
...
+            <h3>{ TRANSLATIONS.completed() }</h3>
+            <p>{ passedExamMessage() }</p>
+            <button class="btn-primary" onclick="{goToAllCourses}">
+                { TRANSLATIONS.back() }
+            </button>
+            <button class="btn-secondary" onclick="{goToResources}">
+                { TRANSLATIONS.resources() }
+            </button>
+        </template>
+        <template if="{ course.examType !== 'prelearning' }">
```
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/Lesson/ExamResults.riot.html]: see [fix exam results](https://github.com/catalpainternational/canoe/pull/598/commits/c7d4be2556e02993aa9d8644e9ee8e5a7b6d4846)

File: src/riot/Lesson/TestPopupModal.riot.html [hunk 7]
```diff
@@ -38,47 +38,70 @@ <h3>{ props.isQuestionComplete ? correctTitleMessage(props.selectedAnswer) : inc
...
             },
 
             completePage() {
-                this.props.toggleModal(false, {});
-                this.props.lesson.completeSection('content')
+                this.props.toggleModal(false);
+                this.props.lesson.complete = true;
+
+                // clear the answers from memory so they do not show up next time
+                clearPageTestAnswers(this.lesson.id);
```
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/Lesson/TestPopupModal.riot.html]: see fix t[est popup modal](https://github.com/catalpainternational/canoe/pull/598/commits/e32838259272d3004a60288e3b4c53c5b64a3edd)

File: src/riot/WagtailPages/LessonPage.riot.html [hunk 8]
```diff
@@ -36,20 +36,14 @@ <h3>{lesson.title}</h3>
...
-            onMounted(props, state) {
-                this.disableBlockedLessonModules();
-            },
-
-            onUpdated() {
-                this.disableBlockedLessonModules();
+            get cardImageUrl() {
+                this.lesson.cardImage
+                    ? `${process.env.API_BASE_URL}${this.lesson.cardImage}`
+                    : undefined;
```
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/WagtailPages/LessonPage.riot.html:46]: see [fix lesson page](https://github.com/catalpainternational/canoe/pull/598/commits/79a5ef013ba117b3ad6065950668faef184936a9)

File: src/riot/Lesson/TestPopupModal.riot.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/Lesson/TestPopupModal.riot.html]: see fix [test popup modal](https://github.com/catalpainternational/canoe/pull/598/commits/e32838259272d3004a60288e3b4c53c5b64a3edd)
Which moves completely away from the props.lesson to just use the page get attribute

File: src/riot/WagtailPages/LessonPage.riot.html [hunk 9]
```diff
@@ -9,7 +9,7 @@ <h3>{lesson.title}</h3>
             <div if="{state.cardImageUrl}" class="lesson-page-image-wrap">
```
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/WagtailPages/LessonPage.riot.html]: see [fix lesson page](https://github.com/catalpainternational/canoe/pull/598/commits/79a5ef013ba117b3ad6065950668faef184936a9)

**Comment by PeteCoward** [comment]
Thanks @megganeturner hugely appreciate your review on this!

**Review by megganeturner — APPROVED** [review]
Nice one Peter! This looks good to me

**Pull request merged**